state/accountsDB.go

```diff
@@ -329,22 +372,31 @@ func (adb *AccountsDB) saveCode(newAcc, oldAcc baseAccountHandler) error {
 		newCodeHash = adb.hasher.Compute(string(newCode))
 	}
 
+	if adb.enableEpochsHandler.IsFlagEnabled(common.MigrateCodeLeafFlag) &&
+		newAcc.GetVersion() == uint8(core.WithoutCodeLeaf) && oldAccVersion == uint8(core.NotSpecified) {
+		return adb.migrateCode(newAcc, oldAccVersion, newAccVersion, newCodeHash, oldCodeHash, newCode)
+	}
+
+	if !newAcc.HasNewCode() {
+		return nil
+	}
+
 	if bytes.Equal(oldCodeHash, newCodeHash) {
 		newAcc.SetCodeHash(newCodeHash)
 		return nil
 	}
 
-	unmodifiedOldCodeEntry, err := adb.updateOldCodeEntry(oldCodeHash)
+	unmodifiedOldCodeEntry, err := adb.updateOldCodeEntry(oldCodeHash, oldAccVersion)
```

Comment:
L389->L405 is the same as migrateCode(), (except `updateNewCodeEntry()` vs `GetStorageManager().Put`). I think you can use `updateNewCodeEntry()` in both situations. Rename `migrateCode()` to something like `updateCodeInStorage()` and use it here and where `migrateCode()` is called.

Reply:
refactored here